### PR TITLE
angularjs: added $q tests

### PR DESCRIPTION
**Pull request by chrootsu**
https://docs.angularjs.org/api/ng/service/$q

### CODE REVIEW

File: angularjs/angular-tests.ts [hunk 1]
```diff
@@ -244,17 +244,44 @@ foo.then((x) => {
...
+    var tResult: TResult;
+    var promiseTResult: ng.IPromise<TResult>;
+
     var $q: ng.IQService;
     var promise1: ng.IPromise<any>;
     var promise2: ng.IPromise<any>;
 
+    // $q constructor
+    result = <ng.IPromise<TResult>>new $q<TResult>((resolve: (value: TResult) => any) => {});
+    result = <ng.IPromise<TResult>>new $q<TResult>((resolve: (value: TResult) => any, reject: (value: any) => any) => {});
```
Comment (vvakame) [inline angularjs/angular-tests.ts]: I think `<ng.IPromise<TResult>>` is not needed.

File: angularjs/angular-tests.ts [hunk 2]
```diff
@@ -244,17 +244,44 @@ foo.then((x) => {
...
 module TestQ {
+    var result: any;
+
+    interface TResult {
+        a: number;
+        b: string;
+        c: boolean;
+    }
+    var tResult: TResult;
+    var promiseTResult: ng.IPromise<TResult>;
```
Comment (vvakame) [inline angularjs/angular-tests.ts]: please use `angular.IPromise` instead of `ng.IPromise`.

File: angularjs/angular-tests.ts [hunk 3]
```diff
@@ -244,17 +244,44 @@ foo.then((x) => {
...
+
     // $q.all
     $q.all([promise1, promise2]).then((results: any[]) => {});
-    $q.all<number>([promise1, promise2]).then((results: number[]) => {});
+    $q.all<TResult>([promise1, promise2]).then((results: TResult[]) => {});
     $q.all({a: promise1, b: promise2}).then((results: {[id: string]: any;}) => {});
     $q.all<{a: number; b: string;}>({a: promise1, b: promise2}).then((results: {a: number; b: string;}) => {});
+
+    // $q.defer
+    result = <ng.IDeferred<TResult>>$q.defer<TResult>();
```
Comment (vvakame) [inline angularjs/angular-tests.ts]: same as above.

**Comment by chrootsu** [comment]
@vvakame Unfortunately, I do not understand why in the case of the constructor does not need to check the return type. Could you explain to me, please?
The rest is fixed.

**Comment by vvakame** [comment]
type assertion is not component of type checking. it is hint for compiler.
I think type inference is more useful than type hint.

http://goo.gl/KwW9Wo

```
// bad. too long and programmer is forced to clever everytime.
var result: any;
result = <angular.IPromise<TResult>>(new $q<TResult>((resolve: (value: TResult) => any) => {}));

// good! short and safe.
{
  let result: angular.IPromise<TResult>;
  result = new $q((resolve: (value: TResult) => any) => {});
}
```

http://www.typescriptlang.org/Handbook#generics
@basarat do you have any other advice?

**Comment by basarat** [comment]
> type assertion is not component of type checking. it is hint for compiler.
> I think type inference is more useful than type hin

:+1: 

The issue with a type assertion is that it checks bivariance in *both* directions which can silently swallow errors : 

```ts
interface Foo {
	foo: number;
}
interface Bar {
	bar: number;
}

var foo:Foo;
foo = {foo:123}; // okay
foo = {}; // Error
foo = <Foo>{}; // Okay as `Foo` is assignable to `{}`
foo = <Foo>{bar:123}; // Error as `Foo` is not assignable to `Bar` and `Bar` is not assignable to `Foo`
```

**Comment by chrootsu** [comment]
@vvakame @basarat Thank you very much, I think I understood ^_^

**Comment by chrootsu** [comment]
@vvakame Fixed.

**Pull request merged**